### PR TITLE
Closes #39 [feature/friendship-system]

### LINKED ISSUE #39: Create friendship relations between users

### PR DESCRIPTION
**Pull request by karimassi**

### CODE REVIEW

File: app/src/androidTest/java/ch/epfl/balelecbud/authentication/MockAuthenticator.java [hunk 1]
```diff
@@ -32,169 +25,75 @@
...
         loggedIn = false;
     }
 
     @Override
-    public void signIn(final String email, final String password, Callback callback) {
+    public CompletableFuture<User> signIn(final String email, final String password) {
         if (users.containsKey(email) && users.get(email).equals(password)) {
             setLoggedIn(true);
-            callback.onSuccess();
+            return MockDatabaseWrapper.getInstance().getDocument("users", "0", User.class).thenCompose(new Function<User, CompletionStage<User>>() {
```
Comment (axelmarmet) [inline app/src/androidTest/java/ch/epfl/balelecbud/authentication/MockAuthenticator.java]: why not use thenApply and hide the CompletionStage ?

File: app/src/androidTest/java/ch/epfl/balelecbud/authentication/MockAuthenticator.java [hunk 2]
```diff
@@ -32,169 +25,75 @@
...
             users.put(email, password);
             setLoggedIn(true);
-            callback.onSuccess();
+            User u = new User(email, email, String.valueOf(uid));
+            uid++;
+            MockDatabaseWrapper.getInstance().storeDocumentWithID("users", String.valueOf(uid), u);
+            return CompletableFuture.completedFuture(null);
         } else {
-            callback.onFailure("Registration failed: account already exists with this email");
+            return CompletableFuture.completedFuture(null).thenApply(new Function<Object, Void>() {
```
Comment (axelmarmet) [inline app/src/androidTest/java/ch/epfl/balelecbud/authentication/MockAuthenticator.java]: Weird that there is no way to create an ExceptionnalFuture but I also looked in the interface and did not find a cleaner way. Might be nice to hide in an util function though ?

File: app/src/androidTest/java/ch/epfl/balelecbud/authentication/MockAuthenticator.java (hunk 1 above)
Comment (karimassi) [reply app/src/androidTest/java/ch/epfl/balelecbud/authentication/MockAuthenticator.java]: You're right, it makes more sense to use thenApply here.

File: app/src/main/java/ch/epfl/balelecbud/models/FriendRequest.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        if (o == null || getClass() != o.getClass()) return false;
+        FriendRequest request = (FriendRequest) o;
+        return getSenderId().equals(request.getSenderId()) &&
+                getRecipientId().equals(request.getRecipientId());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getSenderId(), getRecipientId());
+    }
```
Comment (axelmarmet) [inline app/src/main/java/ch/epfl/balelecbud/models/FriendRequest.java]: Use the actual attributes instead of the getters

Comment (karimassi) [reply app/src/main/java/ch/epfl/balelecbud/models/FriendRequest.java]: Method generated by android studio :)

File: app/src/androidTest/java/ch/epfl/balelecbud/authentication/MockAuthenticator.java (hunk 2 above)
Comment (karimassi) [reply app/src/androidTest/java/ch/epfl/balelecbud/authentication/MockAuthenticator.java]: You're right. A `CompletableFutureUtil` class would be nice to have in order to provide such futures. Thanks !

**Comment by karimassi** [comment]
> Futures are indeed nicer that callback functions ;-)
> Is it normal that the `signIn` function of `MockAuthenticator` sets `currentUser` whereas `FirebaseAuthenticator`'s one doesn't ?
> Couldn't we change `Authenticator` to be an `abstract class` instead of an `interface` ?

No, thanks for pointing that out! I changed it. I think we can keep the Authenticator as an Interface for now.

**Pull request merged**